src/workflow-requests.ts

```diff
@@ -20,79 +20,103 @@ import type {
+  const workflowContextClient = firstInvocationParams[0].workflowContext.qstashClient;
 
-  if (useJSONContent) {
-    headers["content-type"] = "application/json";
-  }
+  const invocationBatch = firstInvocationParams.map((invocation) => {
+    const { headers } = getHeaders({
+      initHeaderValue: "true",
+      workflowConfig: {
+        workflowRunId: invocation.workflowContext.workflowRunId,
+        workflowUrl: invocation.workflowContext.url,
+        failureUrl: invocation.workflowContext.failureUrl,
+        retries: invocation.workflowContext.retries,
+        telemetry: invocation.telemetry,
+        flowControl: invocation.workflowContext.flowControl,
+        useJSONContent: invocation.useJSONContent ?? false,
+      },
+      invokeCount: invocation.invokeCount ?? 0,
+      userHeaders: invocation.workflowContext.headers,
+    });
+
+    // QStash doesn't forward content-type when passed in `upstash-forward-content-type`
+    // so we need to pass it in the headers
+    if (invocation.workflowContext.headers.get("content-type")) {
+      headers["content-type"] = invocation.workflowContext.headers.get("content-type")!;
+    }
+
+    if (invocation.useJSONContent) {
+      headers["content-type"] = "application/json";
+    }
 
-  try {
     const body =
-      typeof workflowContext.requestPayload === "string"
-        ? workflowContext.requestPayload
-        : JSON.stringify(workflowContext.requestPayload);
-    const result = await workflowContext.qstashClient.publish({
+      typeof invocation.workflowContext.requestPayload === "string"
+        ? invocation.workflowContext.requestPayload
+        : JSON.stringify(invocation.workflowContext.requestPayload);
+
+    return {
       headers,
       method: "POST",
       body,
-      url: workflowContext.url,
-      delay,
-    });
+      url: invocation.workflowContext.url,
+      delay: invocation.delay,
+    } as PublishBatchRequest;
+  });
 
-    if (result.deduplicated) {
-      await debug?.log("WARN", "SUBMIT_FIRST_INVOCATION", {
-        message: `Workflow run ${workflowContext.workflowRunId} already exists. A new one isn't created.`,
-        headers,
-        requestPayload: workflowContext.requestPayload,
-        url: workflowContext.url,
-        messageId: result.messageId,
-      });
+  try {
+    const results = (await workflowContextClient.batch(invocationBatch)) as PublishToUrlResponse[];
+
+    const invocationStatuses: ("success" | "workflow-run-already-exists")[] = [];
+
+    for (let i = 0; i < results.length; i++) {
+      const result = results[i];
+      const invocationParams = firstInvocationParams[i];
+      if (result.deduplicated) {
+        await invocationParams.debug?.log("WARN", "SUBMIT_FIRST_INVOCATION", {
+          message: `Workflow run ${invocationParams.workflowContext.workflowRunId} already exists. A new one isn't created.`,
+          headers: invocationParams.workflowContext.headers,
```

Comment:
looking at the failing test, I think there is a change in headers. before it was Record<string, string>, now it is Headers, which doesn't print the headers explicitly.

if we change this to invocationParams.headers, test may pass.

Same below.